### PR TITLE
[MLIR][LLVM] Add import flag to skip traversal of DICompositeType's elems

### PR DESCRIPTION
**Pull request by Dinistro**
This commit introduces a flag to allow skipping the potentially recursive import of DICompositType elements. This patch is essentially a bandaid for the still broken recursive debug type import.

Some of our downstream inputs are produced by excessive usage of template meta programming, and thus contain tens of thousands of types that all participate in such recursions. Unfortunately, the series of patches that introduces type support is not easily revertable due to being around for a while now and Modular depending on it.

We can consider to revert this change once the type importer has show to be very performant, but for now we are talking second vs hours to import specific files.

### CODE REVIEW

File: mlir/lib/Target/LLVMIR/ConvertFromLLVMIR.cpp [hunk 1]
```diff
@@ -30,6 +30,11 @@ void registerFromLLVMIRTranslation() {
       llvm::cl::desc("Emit expensive warnings during LLVM IR import "
                      "(discouraged: testing only!)"),
       llvm::cl::init(false));
+  static llvm::cl::opt<bool> importEmptyDICompositeTypes(
+      "import-empty-di-composite-types",
```
Comment (Dinistro) [inline mlir/lib/Target/LLVMIR/ConvertFromLLVMIR.cpp:34]: We could also call this something like "avoid-recursive-type-import" or something. I'm open to suggestions 😄

Comment (gysit) [reply mlir/lib/Target/LLVMIR/ConvertFromLLVMIR.cpp:34]: Maybe `drop-di-composite-type-elements` or `strip-di-composite-type-elements` is a bit more precise? The current naming works as well.

File: mlir/lib/Target/LLVMIR/DebugImporter.h [hunk 2]
```diff
@@ -184,6 +184,12 @@ class DebugImporter {
 
   MLIRContext *context;
   ModuleOp mlirModule;
+
+  /// An option to control if DICompositeTypes should always be imported without
+  /// converting their elements. This is a way to avoid recursive traversals of
+  /// types, which is currently still flawed for inputs produced by extensive
+  /// usage of template meta programming.
```
Comment (gysit) [inline mlir/lib/Target/LLVMIR/DebugImporter.h:191]:
```suggestion
  /// converting their elements. If set, the option avoids the recursive traversal
  /// of composite type debug information, which can be expensive for adversarial inputs.
```
nit:

File: mlir/include/mlir/Target/LLVMIR/Import.h [hunk 3]
```diff
@@ -34,12 +32,17 @@ class ModuleOp;
...
 /// implementation of the LLVMImportDialectInterface. It returns nullptr if the
 /// translation fails and reports errors using the error handler registered with
-/// the MLIR context. The `emitExpensiveWarnings` option controls if expensive
+/// the MLIR context.
+/// The `emitExpensiveWarnings` option controls if expensive
 /// but uncritical diagnostics should be emitted.
+/// The `importEmptyDICompositeTypes` option controls if DICompositeTypes should
+/// be imported without elements. This is a way to avoid triggering any parts of
+/// the recursive DI metadata import, which can consume substantial amounts of
+/// time.
```
Comment (gysit) [inline mlir/include/mlir/Target/LLVMIR/Import.h:41]:
```suggestion
/// be imported without elements. If set, the option avoids the recursive traversal
/// of composite type debug information, which can be expensive for adversarial inputs.
```

**Review by gysit — APPROVED** [review]
LGTM

I made an optional suggestion for renaming the flag and have one nit comment regarding the flag documentation.

File: mlir/lib/Target/LLVMIR/ConvertFromLLVMIR.cpp (hunk 1 above)
Comment (Dinistro) [reply mlir/lib/Target/LLVMIR/ConvertFromLLVMIR.cpp:34]: Changing to drop-di-composite-type-elements

**Comment by zyx-billy** [comment]
Thank you! We probably should've done this earlier to unblock 🙏.

**Pull request merged**